### PR TITLE
Windows 10/11 Calculator: refactor results announcement with UIA notification event

### LINKED ISSUE #13383: Windows 10/11 Calculator: allow NVDA to announce more operations by suppressing only a limited number of keyboard commands
Hi,

This amounts to a substantial redesign of Windows 10/11 Calculator app module, therefore I think it would be best to queue this for 2022.2 or later:

### Background:
A consequence of Windows as a Service (WaaS) introduced with Windows 10 (2015) is continuous ap updates, specifically apps preinstalled on PC's. One of these is Calculator (calculator.exe), which went through different user interface designs and added features over the years.

One of the biggest changes in Calculator came in 2018. With Windows 10 Version 1709 (Fal Creators Update) adding UIA notification event (part of IUIAutomation5 interface), Calculator app uses notification event to announce calculator display content changes. Prior to this, a combination of name change and live region was used to announce display content. With NVDA taking advantage of UIA notification event in 2018, it became possible for the screen reader to announce display content on supported Calculator app releases.

Then in 2020, NVDA added support for Windows 10 Calculator, which originated from Windows App Essentials add-on. In its early days, the NVDA Core version of Calculator app allowed NVDA to announce display content when a limited number of keyboard keys are pressed such as Escape, Enter, and Delete. Refinements followed such as announcing display content on braille displays, and with 2022.1, Windows 11 Calculator support.

In its original design, Calculator app module does not announce display content when number row keys are pressed. In fact, there are keyboard shortcuts in scientific mode such as "S" for sine that are not announced by NVDA. This stems from the fact that NVDA ignores DisplayUpdated activity Id altogether in order to solve double announcement problem where" speak typed characters" enabled causes NVDA to announce both the entered number and display content (for example, typing "2" results in NVDA saying "2, display is 2"). This design led to a discussion on NVDA Users list (see the reference in the end) where a user brought up the following suggestion.

### Is your feature request related to a problem? Please describe.
When a calculator command such as + (plus) or = (equals) is pressed, NVDA does not announce results (specifically, calculator display content). Further, scientific mode results are not announced when operations such as trigonometric, logarithmic, factorial and other functions are performed. The issue was traced to the original design of Calculator app module (see above).

### Describe the solution you'd like
Redesign the UIA notification event handler so it can ignore commands that should not result in NVDA announcing display content. For now number row keys will be marked as "no calculator results" list. Also, change calculator results script so it can just send gestures when using Windows 10 Version 1709 or later as this script was really optimized for an era without UIA notification event (virtually prior to 2018). See the "additional context" section on justifications.

Specifically:

* Current calculator results script wil be edited to just send the gesture if using Windows 10 Version 1709 or later.
* A new script (tentatively named "noCalculatorResults") will e defined and assigned to number row keys (1 through 0). The job of this script is to enable a flag that will inform UIA notification event handler to suppress announcement of display content.
* A private flag will be set up when number keys are entered and cleared inside the UIA notification event handler.
* UIA notification event handler will be edited to return if no results flag is enabled.

### Describe alternatives you've considered
Define a list of keyboard shortcuts for Calculator that NVDA should announce. Note that this approach is not really future-proof because Microsoft can add or change commands in the future, made a bit easier to predict since Calculator is open-source (hosted on GitHub). Another option is keep Calculator app module in Core as is and let add-ons change the behavior, but that is not ideal for folks who may not elect to use add-ons for any reason (and in case NVDA is downloaded from Microsoft Store on educational devices running Windows 10 in S mode or Windows 11 SE).

### Additional context
Below highlights possible questions and justifications for the proposed solution.

#### Why UIA notification event?
Increasingly, as discussed in a recent Windows Terminal issue that @CodeOfDusk commented, Microsoft is electing to use UIA notification event (prototype in Terminal) to notify assistive technologies about screen content changes. The best example, although not working properly for now, is Voice Access in Windows 11 Insider Preview where UIA notification event is used to communicate microphone status; a working example is page load status in Microsoft Edge.

Calculator is no exception, more so since 2018. Calculator does use UIA notification event to communicate screen content changes. Since the issue at hand is NVDA's UIA notification event handler (from client side), it makes sense to refine Calculator's own notification event handler method.

#### You mention Windows 10 Version 1709 a lot. Why?
This is the feature update that introduced UIA notification event, and from practical point of view, it is the earliest Windows 10 release that the proposed solution is applicable to. However, as explained in the next justification, 1709 will not make the cut.

#### How about older Windows 10 releases?
Older Windows 10 releases are not supported by Microsoft except for mission-critical devices. Since these devices are really meant for specialized uses, a version of Windows 10 Enterprise, aprtly named "Windows 10 Enterprise Long-Term Servicing Branch/Channel" is available. Windows 10 Enterprise LTSB/LTSC does not include modern apps such as Calculator, making the proposed solution not applicable.

#### But wait, you said Version 1709 did not make the cut. Why?
As of time of this issue submission, the oldest supported Windows 10 release (apart from LTSB/LTSC releases) is Version 1909 (November 2019 Update) which is going out of support in May 2022. Therefore, from practical point of view, the earliest version that the proposed solution is applicable is Version 20H2 (October 2020 Update). The latest version of Calculator for Windows 10 is 10.2103.8.0, which supports Version 1803 (April 2018 Update) or later.

#### Will the proposed solution work on Windows 11?
Yes, as Windows 10 and 11 Calculator support comes from the base Calculator app module.

#### Is there a prototype?
Yes - Windows App Essentials February 23rd development snapshot or later includes a prototype of the proposed solution.

#### Why delay the implementation to 2022.2 or later?
NVDA 2022.1 beta is very close at hand - master to beta and beta to master data exchange took place a few days ago, which is a key sign that signifies the imminent release of a beta version for wider testing. Also, since 2022.1 is a compatibility breaking release, more effort should be spent on testing these changes. Also, the proposed solution is in is infancy (prototype just released), it will take weeks to gather additional data, which can then inform the accompanying pull request.

### References:
* Windows Calculator repository: https://github.com/microsoft/calculator
* UIA notification event (Win32): https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/UI/Accessibility/fn.UiaRaiseNotificationEvent.html
* NVDA Users list thread on Windows Calculator issue described: https://nvda.groups.io/g/nvda/topic/problem_with_windows/88946458?p=,,,20,0,0,0::recentpostdate/sticky,,,20,2,0,88946458,previd=1645603799979387364,nextid=1645412580606544441&previd=1645603799979387364&nextid=1645412580606544441

Thanks.

### PR DESCRIPTION
**Pull request by josephsl**
### Link to issue number:
Fixes #13383 

### Summary of the issue:
With the initial version of Windows 10/11 Calculator app module, only a select number of keys were defined to announce results. However it turns out there are commands that should be announced such as equals (=) and scientific calculator commands such as s (sign).

### Description of how this pull request fixes the issue:
Refactor UIA notification event handler with use of a new flag specifying gestures that should not announce results. If this flag is on (for now when number keys are pressed), NVDA will not announce results. Also, existing result announcement commands will not announce results in Calculator version 10.1908 as notification event will announce result.

### Testing strategy:
Manual testing (enter multiple calculations in Calculator, including from standard, scientific, and unit converter modes).

### Known issues with pull request:
On newer Calculator releases, NVDA will no longer announce calculator display content if result keys are pressed multiple times, but this aligns better with Narrator behavior.

### Change log entries:
Bug fixes or changes:
In Windows 10 and 11 Calculator version 10.1908 and later, NVDA will announce results when more commands are pressed such as commands from scientific mode. (#13386 

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

File: source/appModules/calculator.py [hunk 1]
```diff
@@ -112,6 +122,11 @@ def script_calculatorResult(self, gesture):
 		# To prevent double focus announcement, check where we are.
 		focus = api.getFocusObject()
 		gesture.send()
+		# #13383: later Calculator releases use UIA notification event to announce results.
+		# While the exact version that introduced UIA notification event cannot be found easily,
+		# it is definitely used in version 10.1908 and later.
+		if self.productVersion >= "10.1908":
```
Comment (seanbudd) [inline source/appModules/calculator.py:128]: I would suggest against using a string comparison here

**Comment by josephsl** [comment]
Hi, that can be arranged within the next half an hour. Thanks.

**Review by seanbudd — APPROVED** [review]

File: source/appModules/calculator.py [hunk 2]
```diff
@@ -125,3 +141,12 @@ def script_calculatorResult(self, gesture):
 				if isinstance(resultsScreen, UIA) and resultsScreen.UIAElement.cachedClassName == "LandmarkTarget":
 					# And no, do not allow focus to move.
 					queueHandler.queueFunction(queueHandler.eventQueue, ui.message, resultsScreen.firstChild.name)
+
+	# Handle both number row and numpad with num lock on.
+	@scriptHandler.script(
+		gestures=[f"kb:{i}" for i in range(10)]
```
Comment (CyrilleB79) [inline source/appModules/calculator.py:147]: Unfortunately this does not work on all keyboard localization. On French keyboard, the digits on the number row (on the alpha-numeric keyboard) are accessible in combination with the shift key.
For example, if you press the key above the 'E' (a bit on the right) you get the open parenthesis character. If you press shift in compination with the key above the 'E', you get the digit '4'.

Also to be noted: some characters on the number row are commands of the calculator (parenthesis, minus sign).
The characters that are on the keys 1 to 0 on the number row of the French keyboard are the following:
`&é"'(-è_çà`

**Review by CyrilleB79 — CHANGES_REQUESTED** [review]

**Comment by josephsl** [comment]
Hi, I bet this may pose a problem for other keyboard layouts as well. Changing to providing a list of Calculator commands that should announce results will also pose a problem if the same thing with keyboard layout happens (that is, Calculator commands differ between keyboard layouts). Using typed character event will not work because you can tab between Calculator UI elements, and if you perform commands while not focused on display content, typed character event will not be invoked (if one does create an overlay class for results display). The best(and ideal) solution is look for a way to detect specific characters that users type throughout an app module; if not, either we need to let translators translate gestures, define an overlay class that will cover the entire Calculator interface (thereby allowing typed character event to function properly), live with gestures defined in this pull request, or say goodbye to Calculator app module altogether and let NVDA repeat things when speak typed characters is enabled and numbers are pressed. I’m also looking for creative solutions that are not on this list of alternatives. Thanks.

**Comment by CyrilleB79** [comment]
I think that this PR and the calculator appModule are worth to be kept, even in case where it would not handle all use cases since it is already an improvement when typing number on the numeric keypad.

I prefer the following solutions if applicable, from better to worse:


Solution 1:
Define a global overlay class to handle typed character event.
This seems good from a user point of view. Maybe harder to read from a dev point of view.

Solution 3: Add more keys to the one causing UIA notifications to be ignored, i. e. kb:number and kb:shift+number
Issue: This cause the result not to be read even if pressing a symbol on the numeric row, e.g. minus symbol on the French keyboard.

Solution 3: Make the alpha-numeric number key translatable.
  Issue: this works only when the keyboard main language matches NVDA's UI language.
Similar assumptions are already present in NVDA's code base, e.g. assuming that Word's UI language matches NVDA's UI one for various Word formatting shortcuts. But this is less than ideal, since one may use two different languages.
As an example for calculator, I am not sure, but I would not be surprised that Canadian people using French NVDA use a keyboard layout which allows to type the number without the shift key.

**Comment by josephsl** [comment]
Hi, a subtle form of global overlay class (which I did try implement hours ago) to address this scenario would be a caching mechanism that will tell NVDA to remember the last gesture or character entered so app modules, global plugins, NVDA objects, and friends can take action. This amounts to a new issue/PR pair and API addition/possible changes. Unfortunately, it doesn’t scale quite well with alternative input systems such as braille display hardware, touch keyboard, voice input, among others. I think for now the way forward would be keep the PR as is and return to the keyboard layout problem when we gather additional data in the future. Thanks.

**Comment by CyrilleB79** [comment]
OK. Thanks for the investigations.

Another option to fix this issue with any type of input device and layout is to report the issue to Microsoft. Since the calculator with Narrator is also very verbose.

**Pull request merged**